Makes IL and EL idempotent

KunJeong opened the pull request:
This PR makes the IL of p4cherry idempotent with EL. This means that the result of typechecking is printed as a **syntactically valid** P4 program **equivalent to the original**. In addition, type information inferred by the typechecker is added whenever possible, including making implicit type casts explicit, and specifying omitted type arguments.
Therefore changes to the codebase consist of the following tasks.
	1. Making the EL -> IL structure reversible
	2. Adjusting the pretty-printing to be syntactically valid
	3. Changes to the binary / test scripts to add roundtrip functionality

1. **Equivalence**: Making the EL -> IL structure reversible
   While most of the typechecker logic does not lose data from the original source code (parsed as EL), there were a few information losses that required dealing with.
	1. Eager evaluation of LCTK expressions
		Static evaluation is performed in several parts of the typechecker to deal with LCTK(local compile-time known). Some expression types which are trivially LCTKs are evaluated into `ValueE`s (namely: `BoolE`, `StrE`, `NumE`, `InvalidE`, `ErrAccE`, `TypeAccE`, and default values for parameters), and the resulting IL loses equivalence with the source EL.
		To solve this problem, the AST is modified as such.
		- a `note` of type `expr` is added to the `value` type of IL AST, and every static evaluation stores the original expression in this note. (`eval_expr` in `typing/static.ml` )
			- in addition, anywhere ValueE is created, the source expression is added as a note
				- affected: `exec/driver.ml`, `instance/instantiate.ml`
			- because the IL and EL AST inherit structure from Lang AST, the `value` type of `lang/ast.ml` is changed to use `note_phrase` instead of `phrase`. EL AST `value` takes a note of type `unit`. 
			- for `walk`, two different `note` type parameters (`'note_value`, `'note_expr`) are passed into `walker` instead of one.
				- affected: `lang/walk.ml` `el/walk.ml`, `il/walk.ml`
		- `il/pp_to_el.ml` uses `value.note` (=the source expression) when printing values (`pp_value`)
		- `BoolE`, `StrE`, `NumE`, `InvalidE`, `ErrAccE`, `TypeAccE` is added to the IL AST.
			- affected: `il/eq.ml`, `il/walk.ml`, `il/pp.ml`
			- corresponding evaluation logic is moved out of `typing/typecheck.ml` and added into `instance/instantiate.ml` and `typing/static.ml` (`eval_num_expr`, `eval_error_acc_expr`, `eval_type_acc_expr`)
		- the type signature of default values in `param` is changed to `expr` instead of `value`.
			- the original structure had a `Runtime-static.Tdomain.Type` module that was used to create the IL AST via dependency injection. However, with the new structure `Type.param` needed access to `expr` which was only present in the instantiated IL AST. Thus the module is no longer injected, instead it is merged into the IL AST and references `expr` via mutual recursion.
				- this necessitated a massive refactoring:
					- (edit [bc12642](https://github.com/kaist-plrg/p4cherry/pull/14/commits/bc12642d8a8b0043fb588e9653abec9726e90c07)) `runtime-static/tdomain` is moved to `runtime-type`, and depends on `il`. 
					- `runtime-static/ctk.ml` is moved to `il/ctk.ml`
					- `runtime-static/vdomain` is moved out and renamed `runtime-value`
					- `runtime-static/builtins`, `dir`, `env.ml`, `numerics.ml` remain in `runtime-static`
					- (edit [bc12642](https://github.com/kaist-plrg/p4cherry/pull/14/commits/bc12642d8a8b0043fb588e9653abec9726e90c07)) `tdomain/pp.ml` is merged into `il/pp.ml` and structural equivalence in `tdomain/eq.ml` is merged into `il/eq.ml`. `eq_alpha` and `eq_kind` are moved into `runtime-type/eq_alpha.ml`.
					- `il/tdomain/tdom.ml` is merged into `il/ast.ml`. `Types.typ`, `Types.param`, `Types.cparam` are now `Il.Ast.typ'`, `Il.Ast.param'` and `Il.Ast.cparam'`. `param'(=cparam')` also has an added `anno`(annotation) field. 
						- affected: `il/tdomain/eq_typ.ml`,  `il/tdomain/pp_typ.ml`,`il/tdomain/free.ml`, `il/tdomain/subst.ml`, `instance/instantiate.ml`, `typing/typecheck.ml`, `typing/wellformed.ml`
	2. Storing ids for structured types (`DefT`, `ParserT`, `ControlT`, `PackageT`)
	   The original type names are needed to print the above types. `id` fields are added into the IL AST for each of these.
		- affected:  `il/tdomain/eq_typ.ml`,  `il/tdomain/pp_typ.ml`,`il/tdomain/free.ml`, `il/tdomain/subst.ml`, `instance/instantiate.ml`, `typing/typecheck.ml`, `typing/wellformed.ml`
	3. Handling polymorphism of SeqT
	    `SeqT` is an internal type used to handle `SeqE` expressions (syntax: `{ a, b }`). It is then cast into `ListT`, `TupleT`, `StackT`, `StructT` or `HeaderT`  based on the left-hand side. This polymorphism is handled internally using type specialization.
	    Expression `{ a, b }` of `type` where `ta` and `tb` are type of `a`, `b` respectively, is converted into `(type<T1,T2><T1,T2>)<ta,tb>` where `type` is any one of `ListT`, `TupleT`, `StackT`, `StructT` and `HeaderT`.
	    - `SeqT` is converted into `TupleT` after type inference is completed. This prevents `SeqT` from appearing in the final IL AST. (`infer_targs` in `typing/typecheck.ml`)
	    - `SpecT` where the inner types are any one of `ListT`, `TupleT` and `StackT` are "canonicalized" (`pp_typ'` in `il/pp_to_el.ml`). Note that we cannot use the actual `Type.canon` function because this would unwrap inner structured types as well.

2. **Validity**: Adjusting the pretty-printing to be syntactically valid
   Most of this is done in `il/pp_to_el.ml`. Many functions are identical to `il/pp.ml`, with the following additions:
	- `pp_typ'` to handle syntactically valid printing of types, including dropping the inner types and printing just the `id`s for structured types.
	- `pp_value'` to do the same for values. Minor adjustments are made.
   - The following  information are dropped due to syntax restrictions:
	   - explicit typecasting to `SetT` is dropped in select expressions (pattern `CastE` in `pp_expr'`), because `set` types are internal and cannot be printed as valid syntax.
		   - **the inferred set types are printed as leading inline comments.**
		     ```p4
		     //example: chain1.p4
		     bit<1> x;
		     state start {
		       transition select(x) {
		         (/* set<bit<1>> */ 0): chain1;
		         (/* set<bit<1>> */ 1): chain2;
		       }
		     }
	   - table entry priorities are not allowed when the table is defined using `const entries` (`pp_table_entry'`)
		   - *"assigning numeric priorities to entries of a table that has`const entries` is not allowed"*, 14.2.1.4 of the p416 spec
		   - **the inferred priorities are printed as leading inline comments.**
		     ```p4
		     //example: table-entries-optional-bmv2.p4
		     const entries = {
		       /* priority = 4 : */ (((bit<8>) (170)), ((bit<16>) (4369))) : a_with_control_params(((bit<9>) (1))); /* data: (bit<9> x) */
		       /* priority = 3 : */ (_, ((bit<16>) (4369))) : a_with_control_params(((bit<9>) (2))); /* data: (bit<9> x) */
		       /* priority = 2 : */ (((bit<8>) (170)), _) : a_with_control_params(((bit<9>) (3))); /* data: (bit<9> x) */
		       /* priority = 1 : */ (_, _) : a_with_control_params(((bit<9>) (4))); /* data: (bit<9> x) */
		     }
		     ```
	   - external method prototype constructors (`ExternConsM` in p4cherry) do not allow type parameters in p4. Don't care types (`AnyT` in p4cherry) in the constructor parameters are used as a workaround to provide polymorphism. Internally, the `AnyT`s are interpreted as if there were the same number of "hidden" type parameters in the constructor (`tparams_hidden`). For syntactic validity, these "hidden" type parameters are removed and substituted back to underscores in constructor parameters during post-processing (`typing/postprocess.ml`)
	     - (edit [5934014](https://github.com/kaist-plrg/p4cherry/pull/14/commits/593401430c750838d83571060a20a331c8732d5d)) The constructor types used in instantiation of these constructors are also substituted similarly (`hide_tparams_constyp` in `typing/postprocess.ml`)
	     - the same needs to be done for type arguments during instantiation. the `targs` of `InstE` and `InstD` in the IL AST are split into `targs` and `targs_hidden` to separate the original type arguments from the inferred.
		     - affected:  `il/eq.ml`,  `il/pp.ml`, `il/walk.ml`, `instance/instantiate.ml`, `typing/typecheck.ml`
		     - type arguments inferred during specialization is separated as `targs_hidden` (`type_instantiation` in `typing/typecheck.ml`)
	     - the same hiding and substitution is performed for `PackageTypeD`. This is to match type signatures with instantiation (although type parameters are allowed in this declaration) (`typing/postprocess.ml`)
	     - **the number of created/matched "don't care types" are printed as trailing comments.**
	       ```p4
	       //example: unused.p4
	       package top(cproto<_> _c); /* 1 wildcards created */
	       top(c()) main; /* 1 wildcards matched */ 
	       ```
	     - **NOTE**: other object and function declarations (`ParserTypeD`, `ControlTypeD`, `ExternFuncD`, etc.) are kept as is, meaning inferred "hidden" type parameters may appear in the left-hand side of these declarations. Might want to change this behavior to match constructors in the future.

3. **Executing / Testing**: 
   - the `-r` flag for command `typecheck` executes the complete roundtrip (input > typecheck > postprocess > AST1, AST1 > print > typecheck > post process > AST2), compares AST1 and AST2, then prints AST2 as output. (`bin/main.ml`)
   - roundtrip tests are performed by default for each positive typecheck test case (`test/test.ml`)

jaehyun1ee commented on p4/lib/typing/tblctx.ml:
```diff
@@ -22,7 +22,7 @@ type mode = NoPri | NoPriLpm of int | Pri | PriLpm
 type t = {
   keys : (Type.t * Il.Ast.match_kind') list;
-  actions : (Il.Ast.var' * Types.param list * Il.Ast.arg list) list;
+  actions : (Il.Ast.var' * Types.param' list * Il.Ast.arg list) list;
```
Let's put `Il.Ast.param'` instead.

KunJeong replied:
addressed in [0377fa7](https://github.com/kaist-plrg/p4cherry/pull/14/commits/0377fa7327356714b6b9304d1e785ff55cd348be)

jaehyun1ee commented on p4/lib/typing/wellformed.ml:
```diff
@@ -541,31 +543,31 @@ and check_valid_typdef'' (tset : TIdSet.t) (td : TypeDef.t) : unit =
    as an argument for such a parameter *)
 
-and check_valid_param (cursor : Ctx.cursor) (ctx : Ctx.t) (param : Types.param)
+and check_valid_param (cursor : Ctx.cursor) (ctx : Ctx.t) (param : Types.param')
```
Let's put `Il.Ast.param'` instead of `Types.param'` and same for elsewhere.

KunJeong replied:
addressed in [0377fa7](https://github.com/kaist-plrg/p4cherry/pull/14/commits/0377fa7327356714b6b9304d1e785ff55cd348be)

jaehyun1ee commented on p4/lib/typing/typecheck.ml:
```diff
@@ -170,22 +170,26 @@ let rec gen_cstr (cstr : cstr_t) (typ_param : Type.t) (typ_arg : Type.t) :
-  | ControlT params_param, ControlT params_arg ->
+  | ParserT (_, params_param), ParserT (_, params_arg)
+  | ControlT (_, params_param), ControlT (_, params_arg) ->
+      (* might need to use id_param and id_arg here *)
```
Parser and control types are structural at type unification. So seems like the code is doing the right thing. Please remove the comment.

jaehyun1ee commented on p4/lib/typing/typecheck.ml:
```diff
@@ -547,7 +565,7 @@ and check_table_apply_as_arg ~(action : bool) (args_il : Il.Ast.arg list) : unit
 
 and type_call_convention ~(action : bool) (cursor : Ctx.cursor) (ctx : Ctx.t)
-    (params : Types.param list) (args_il_typed : (Il.Ast.arg * Type.t) list) :
+    (params : Types.param' list) (args_il_typed : (Il.Ast.arg * Type.t) list) :
```
Let's put `Il.Ast.param'` here instead. Similarly elsewhere.

KunJeong replied:
addressed in [0377fa7](https://github.com/kaist-plrg/p4cherry/pull/14/commits/0377fa7327356714b6b9304d1e785ff55cd348be)

jaehyun1ee commented on p4/lib/typing/subtyp.ml:
```diff
@@ -49,6 +49,13 @@ let check_explicit_castable (typ : Type.t) : bool =
   | _ -> false
 
 let rec explicit (typ_from : Type.t) (typ_to : Type.t) : bool =
+  let typ_from = Type.canon typ_from in
+  let typ_to = Type.canon typ_to in
+  if Type.eq_alpha typ_from typ_to then
+    if check_explicit_castable typ_to then explicit' typ_from typ_to else false
+  else explicit' typ_from typ_to
```
I think this is quite redundant, regarding we have a very similar logic in L188-191. Can't we do something like:

```ocaml
let rec explicit (typ_from : Type.t) (typ_to : Type.t) : bool =
   check_explicit_castable typ_to && explicit' typ_from typ_to
```

KunJeong replied:
addressed in [c7d945b](https://github.com/kaist-plrg/p4cherry/pull/14/commits/c7d945bc116b40887df65b021adf1ebcc13b33f3), we no longer need the inner `explicit'`

jaehyun1ee commented on p4/lib/il/ast.ml:
```diff
@@ -54,11 +51,79 @@ and dir' = L.dir'
+  (* 3b. Data types *)
+  | NewT of L.id' * typ'
+  | EnumT of L.id' * L.member' list
+  | SEnumT of L.id' * typ' * (L.member' * Value.t) list
```
Why not have `SEnumT of L.id' * typ' * (L.member' * value) list` here?

KunJeong replied:
Changed to `value'` instead of `Value.t` : [25b64d1](https://github.com/kaist-plrg/p4cherry/pull/14/commits/25b64d1569caf7ecdd6a50453054ae86cd68b9a4)

jaehyun1ee reviewed (changes requested):
Thank you for the massive amount of refactoring and the detailed description. I've made minor comments, mostly nits.

Macro-structural issues:

- Please order module imports. In general: `module Runtime_value`, followed by `module Il`, followed by `module Runtime_static`, which makes their dependency explicit.
- Qualified subdirs for `il/tdomain`. Although `unqualified` makes imports less wordy, qualified subdirs make the project structure more explicit at code level.
-

KunJeong commented on p4/lib/il/eq.ml:
```diff
@@ -104,13 +107,13 @@ and eq_annos ?(dbg = false) annos_a annos_b =
 (* Type parameters *)
 
-and eq_tparam' tparam_a tparam_b = E.eq_tparam' tparam_a tparam_b
+and eq_tparam' tparam_a tparam_b = Eq_typ.eq_tid' tparam_a tparam_b
```
This concerns the test `p4/testdata/p4c/program/well-typed/unused.p4`. From the following p4 code:
```c++
// original code
extern E {
  E();
  bit<32> get<T>(in T b);
}
control c(inout S s)() {
  E() e;
  apply {
    if (s.h.isValid()) s.h.data3 = 0;
    if (((s.h.data2) == (0))) s.h.data1 = e.get(s.h.data2); //__WILD_0
  }
}
control cproto<T>(inout T v);
package top(cproto<_> _c); //__WILD_1
top(c()) main;

```
Two fresh type ids are inserted as indicated in the comments, one in the "get" method call and one in the wildcard.


However, if we were to typecheck-and-print this code, it would add an inferred type argument to the "get" method call. (explicit casting is also added, but omitted here as it is irrelevant) Therefore, the wildcard is now the only fresh type id, which makes it "__WILD_0" instead of "__WILD_1".
```c++
// typechecked code
...
  apply {
    if (s.h.isValid()) s.h.data3 = 0;
    if (((s.h.data2) == (0))) s.h.data1 = e.get<bit<32>>(s.h.data2);
  }
}
control cproto<T>(inout T v);
package top(cproto<_> _c); //__WILD_0
top(c()) main;
```
Therefore the two ASTs are equivalent except for the 'tick' count, which is why I special-cased "__WILD" tids. (The alternative was to drop the inferred type argument, but it felt less ideal.)
However, I agree that it is indeed a bit of a hack. Perhaps I could use alpha-equivalence somewhere to get around this issue, but it is tricky as these are not really free type variables. Let me know what you think.

KunJeong replied:
Addressed in [5934014](https://github.com/kaist-plrg/p4cherry/pull/14/commits/593401430c750838d83571060a20a331c8732d5d). Special handling of `WILD` in equivalence is removed, and instead `WILD` VarTs are substituted out of constructors instantiations (`InstD` and `CallInstS`) in logic similar to their constructors (`PackageTypeD` and `ExternConsM`) in `typing/postprocess.ml`.

KunJeong commented on p4/lib/il/dune:
```diff
@@ -1,4 +1,8 @@
+(include_subdirs unqualified)
```
`qualified` is definitely doable, but it does not solve the renaming issue - for now, calling module `Pp` in `Tdomain/subst.ml` is resolved to `Il.Pp` by dune, and I haven't yet figured out a way to call `Tdomain.Pp` which is what we want. Seems like an issue with dune, still searching for relevant issues in the repo.
Also, every submodule in `tdomain` imports `Il.Ast`, so the current directory structure does not accurately reflect the module dependencies. This is another reason I opted for `qualified` (trying to represent a "flattened" structure), and I think a better way to go about this might just be to join `pp` and `pp_typ` into one file and similarly for`eq`/`eq_typ` (no longer makes sense to keep them in separate files)

KunJeong replied:
Addressed in [bc12642](https://github.com/kaist-plrg/p4cherry/pull/14/commits/bc12642d8a8b0043fb588e9653abec9726e90c07) as `tdomain` is moved out of `il`. Will leave this open until the refactoring is documented in the PR

KunJeong replied:
Added to PR documentation.

KunJeong commented on p4/lib/il/eq.ml:
```diff
@@ -1,11 +1,14 @@
 module E = Lang.Eq
-module Ctk = Runtime_static.Ctk
-module Types = Runtime_static.Tdomain.Types
+module Ctk = Ctk
```
You are right. Removed in [57208e3](https://github.com/kaist-plrg/p4cherry/pull/14/commits/57208e303b153c3904a15b86176db6ea391cdd8c).

KunJeong commented on p4/lib/il/pp_to_el.ml:
```diff
@@ -0,0 +1,665 @@
+        (pp_pairs ~level:(level + 1) pp_member' pp_value' ~rel:Eq
+           ~sep:SemicolonNl)
+        fields (indent level)
+  | _ -> Value.pp ~level fmt value
```
You are right, this is redundant. Removed in [57208e3](https://github.com/kaist-plrg/p4cherry/pull/14/commits/57208e303b153c3904a15b86176db6ea391cdd8c).

KunJeong commented on p4/lib/instance/instantiate.ml:
```diff
@@ -1005,7 +1050,20 @@ and eval_table_decl (cursor : Ctx.cursor) (ctx : Ctx.t) (sto : Sto.t) (id : id)
   let decl =
     let expr_inst =
-      ValueE { value = value $ no_info }
+      (* unused note *)
```
Removed in [57208e3](https://github.com/kaist-plrg/p4cherry/pull/14/commits/57208e303b153c3904a15b86176db6ea391cdd8c).

KunJeong commented on p4/lib/instance/instantiate.ml:
```diff
@@ -964,12 +1012,9 @@ and eval_parser_type_decl (cursor : Ctx.cursor) (ctx : Ctx.t) (id : id)
-    in
-    let typ_parser = Types.ParserT params in
+    let params = List.map it params in
+    (* unsure if the id here is id *)
+    let typ_parser = Types.ParserT (id.it, params) in
```
Removed in [57208e3](https://github.com/kaist-plrg/p4cherry/pull/14/commits/57208e303b153c3904a15b86176db6ea391cdd8c)

KunJeong replied:
Removed in [f66f6f9](https://github.com/kaist-plrg/p4cherry/pull/14/commits/f66f6f901d8b75fbeff15208bc7c164f7095e4bf).

KunJeong commented on p4/lib/il/eq.ml:
```diff
@@ -1,11 +1,14 @@
 module SType = Envs.SType
 module Frame = Envs.Frame
 module TDEnv = Envs.TDEnv
+*)
```
Removed in [7b3704a](https://github.com/kaist-plrg/p4cherry/pull/14/commits/7b3704a151dd4eae533b7142ff4c13c7cb39fa3d).

KunJeong commented on p4/lib/il/ast.ml:
```diff
@@ -54,11 +51,79 @@ and dir' = L.dir'
+  | SpecT of typdef_poly * typ' list
+  (* 3. Defined types *)
+  (* 3a. Alias types *)
+  | DefT of typ' * L.id'
```
Refactoring done: [25b64d1](https://github.com/kaist-plrg/p4cherry/pull/14/commits/25b64d1569caf7ecdd6a50453054ae86cd68b9a4).

KunJeong commented on p4/lib/il/ast.ml:
```diff
@@ -354,5 +431,13 @@ and mthd' =
+(* Constructor definitions *)
+
+type consdef = tparam' list * tparam' list * constyp
+
```
Addressed in [e277496](https://github.com/kaist-plrg/p4cherry/pull/14/commits/e2774967cf259d9100d10e51c3d8d6a9adb34c3d). `il/pp.ml` and `il/eq.ml` are re-ordered to mimic Ast order as well.

KunJeong commented on p4/lib/il/pp_to_el.ml:
```diff
@@ -0,0 +1,665 @@
+      F.fprintf fmt "{\n%a\n%s}"
+        (pp_pairs ~level:(level + 1) pp_member' pp_value' ~rel:Eq
+           ~sep:SemicolonNl)
+        fields (indent level)
```
This was unreachable code, removed in [57208e3](https://github.com/kaist-plrg/p4cherry/pull/14/commits/57208e303b153c3904a15b86176db6ea391cdd8c).

KunJeong commented on p4/lib/instance/instantiate.ml:
```diff
@@ -1005,7 +1050,20 @@ and eval_table_decl (cursor : Ctx.cursor) (ctx : Ctx.t) (sto : Sto.t) (id : id)
+            $$ ( no_info,
+                 InstE
+                   {
+                     var_inst = L.Top id $ no_info;
```
Thanks for the correction. Addressed in [054fed5](https://github.com/kaist-plrg/p4cherry/pull/14/commits/054fed516c707fdcc078415eed011dda697d5cb1).

KunJeong commented on p4/lib/instance/instantiate.ml:
```diff
@@ -1,10 +1,11 @@
+module TypeDef = Types.TypeDef
 module Numerics = Runtime_static.Numerics
 module Builtins = Runtime_static.Builtins
 module Envs_static = Runtime_static.Envs
```
Reordering done, [054fed5](https://github.com/kaist-plrg/p4cherry/pull/14/commits/054fed516c707fdcc078415eed011dda697d5cb1)

KunJeong commented on p4/lib/il/pp_to_el.ml:
```diff
@@ -0,0 +1,665 @@
+  | AnyT -> F.fprintf fmt "_"
+  | TableEnumT _ | TableStructT _ | SeqT _ | SeqDefaultT _ | RecordT _
+  | RecordDefaultT _ | DefaultT | InvalidT | SetT _ | StateT ->
+      Type.pp ~level fmt typ
```
Addressed in [054fed5](https://github.com/kaist-plrg/p4cherry/pull/14/commits/054fed516c707fdcc078415eed011dda697d5cb1).

KunJeong commented on p4/lib/il/pp_to_el.ml:
```diff
@@ -0,0 +1,665 @@
+      F.fprintf fmt "%a" pp_id' id
+  | ListT _ | TupleT _ -> Type.pp ~level fmt typ
+  | StackT (typ, size) ->
+      F.fprintf fmt "%a[%a]" (pp_typ' ~level:(level + 1)) typ Bigint.pp size
```
Indeed, these are unreachable. Addressed in [054fed5](https://github.com/kaist-plrg/p4cherry/pull/14/commits/054fed516c707fdcc078415eed011dda697d5cb1).

KunJeong commented on p4/lib/il/tdomain/eq_typ.ml:
```diff
@@ -25,9 +31,16 @@ and eq_params params_a params_b = E.eq_list eq_param params_a params_b
+and eq_tid' id_a id_b =
+  String.starts_with ~prefix:"__WILD_" id_a
+  && String.starts_with ~prefix:"__WILD_" id_b
+  || id_a = id_b
```
Closing in favor of https://github.com/kaist-plrg/p4cherry/pull/14#discussion_r2030065872

KunJeong commented on p4/lib/el/pp.ml:
```diff
@@ -484,24 +484,28 @@ and pp_table_actions ?(level = 0) fmt table_actions =
   F.fprintf fmt "%s%s%a%s%a : %a"
     (if table_entry_const then "const " else "")
-    (if table_entry_priority |> Option.is_some then "priority = " else "")
+    (if (not table_entries_const) && table_entry_priority |> Option.is_some then
```
change to ignore `table_entries_const` in [054fed5](https://github.com/kaist-plrg/p4cherry/pull/14/commits/054fed516c707fdcc078415eed011dda697d5cb1).

KunJeong commented:
Thank you for the detailed review! I addressed all issues, and updated the original comment with edits and links to corresponding commits. Here's a more detailed description of what's changed:

Cleaning up the module structure and dependency graph (completed [bc12642](https://github.com/kaist-plrg/p4cherry/pull/14/commits/bc12642d8a8b0043fb588e9653abec9726e90c07))
- `runtime-static/tdomain` is moved to `runtime-type`, and depends on `il`.
- `tdomain/pp.ml` is merged into `il/pp.ml` and structural equivalence in `tdomain/eq.ml` is merged into `il/eq.ml`. `eq_alpha` and `eq_kind` are moved into `runtime-type/eq_alpha.ml`.
- the current dependency graph is ... `runtime-value` > `il` > `runtime-type` > `runtime-static` > ... . We discussed making `runtime-value` symmetric with `runtime-type` (by merging the ast, pp, eq to IL and creating `Value` module after `IL`) as a future task.

Post-processing using transform walker (completed [5934014](https://github.com/kaist-plrg/p4cherry/pull/14/commits/593401430c750838d83571060a20a331c8732d5d))
- the logic needed for substituting `__WILD` (inferred hidden type parameters) out is moved to a separate step, `typing/postprocess.ml`. As a result, the roundtrip cycle is now typecheck > postprocess > print > typecheck > postprocess. 
- since `__WILD` type variables are substituted out of constructors, they need to be removed from the inferred constructor types when instantiating the constructors. This logic is added in postprocessing, this removes the need for special-handling `__WILD_a` and `__WILD_b` in structural equivalence.

Fix: casting for nestable types (completed [c7d945b](https://github.com/kaist-plrg/p4cherry/pull/14/commits/c7d945bc116b40887df65b021adf1ebcc13b33f3))
- p4cherry interprets { 1, "a" } as a `SeqT` and later explicit-casts the `SeqT` to a Tuple. If we print and typecheck this program again, we would see a tuple being explicit-cast to itself. To make this pass the typecheck, we allow all ["nestable types"](https://p4.org/p4-spec/docs/p4-16-working-draft.html#sec-type-nesting) to be cast to themselves (`typing/subtyp.ml`) and add evaluation rules for them (`exec/interp.ml`)

jaehyun1ee commented:
Thanks!

> NOTE: other object and function declarations (ParserTypeD, ControlTypeD, ExternFuncD, etc.) are kept as is, meaning inferred "hidden" type parameters may appear in the left-hand side of these declarations.

One quick catch about the part above. Our decision was to remove the "hidden" type arguments for `InstD` and `CallInstS`, so that would also include the instantiation of parsers and controls. So shouldn't we also remove the type parameters at `ParserD` and `ControlD` as well, similar to `PackageTypeD`?

`ParserTypeD` and `ControlTypeD` wouldn't matter since we don't instantiate them, they are more like type declarations rather than constructor declarations.

e.g.,

```p4
// def-site: typecheck->postprocess would keep the hidden type param for "_"
parser P(in bool x)(bool y, _ z) { apply {} }

// use-site: typecheck->postprocess would drop the hidden type arg
P.apply();
```

jaehyun1ee reviewed (commented):
Last-minute question before merging!

jaehyun1ee commented on p4/bin/main.ml:
```diff
@@ -36,11 +52,17 @@ let typecheck_command =
+           let func = if roundtrip_flag then roundtrip_il else typecheck in
+           func includes filename
+         in
+         program |> Typing.Postprocess.postprocess_program
```
Please remove the redundant postprocessing before merge.

I think it would be better to print the typechecked program as-is, without postprocessing for the `typecheck` command. One consequence of moving `Typing.Postprocess.postprocess_program` might be: the interpreter breaking because it fails to recognize the deleted type params.

jaehyun1ee reviewed (approved):
LGTM! Thanks again for the hard work, and sorry for the late review :)
